bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/EntryLogger.java

```diff
@@ -327,54 +333,100 @@ private int readFromLogChannel(long entryLogId, BufferedReadChannel channel, Byt
         }
     };
 
+    private final  CacheLoader<Long, BufferedReadChannel> readChannelLoader =
+            new CacheLoader<Long, BufferedReadChannel> () {
+        public BufferedReadChannel load(Long entryLogId) throws Exception {
+
+            return getChannelForLogId(entryLogId);
+
+        }
+    };
+
+
+    private static class ReferenceCountedFileChannel extends AbstractReferenceCounted {
+        private final FileChannel fc;
+
+        public ReferenceCountedFileChannel(FileChannel fileChannel) {
+            this.fc = fileChannel;
+        }
+
+
+        @Override
+        public ReferenceCounted touch(Object hint) {
+            return this;
+        }
+
+        // when the refCnt decreased to
+        @Override
+        protected void deallocate() {
+            try {
+                fc.close();
+            } catch (IOException e) {
+                LOG.warn("Exception occurred in ReferenceCountedFileChannel"
+                        + " while closing channel for log file: {}", fc);
+            } finally {
+                IOUtils.close(LOG, fc);
+            }
+        }
+
+    }
+
+
     /**
      * Each thread local buffered read channel can share the same file handle because reads are not relative
-     * and don't cause a change in the channel's position. We use this map to store the file channels. Each
-     * file channel is mapped to a log id which represents an open log file.
+     * and don't cause a change in the channel's position.
+     * Each file channel is mapped to a log id which represents an open log file.
+     * when ReferenceCountedFileChannel's refCnt decrease to 0, close the fileChannel.
      */
-    private final ConcurrentMap<Long, FileChannel> logid2FileChannel = new ConcurrentHashMap<Long, FileChannel>();
+    private ConcurrentMap<Long, ReferenceCountedFileChannel>
+            logid2FileChannel = new ConcurrentHashMap<>();
+
 
     /**
      * Put the logId, bc pair in the map responsible for the current thread.
      * @param logId
      * @param bc
      */
-    public BufferedReadChannel putInReadChannels(long logId, BufferedReadChannel bc) {
-        Map<Long, BufferedReadChannel> threadMap = logid2Channel.get();
-        return threadMap.put(logId, bc);
+    public void putInReadChannels(long logId, BufferedReadChannel bc) {
+        Cache<Long, BufferedReadChannel> threadCahe = logid2Channel.get();
+        threadCahe.put(logId, bc);
     }
 
     /**
      * Remove all entries for this log file in each thread's cache.
      * @param logId
      */
     public void removeFromChannelsAndClose(long logId) {
-        FileChannel fileChannel = logid2FileChannel.remove(logId);
-        if (null != fileChannel) {
-            try {
-                fileChannel.close();
-            } catch (IOException e) {
-                LOG.warn("Exception while closing channel for log file:" + logId);
-            }
-        }
+        //remove the fileChannel from logId2Channel
```

Comment:
I don't think you need to invalidate the local thread cache. The original behavior is to close the file channel directly, the local thread cache might be still referencing this file channel, but it is okay since the file channel is anyway closed.

I would suggest

- add a method called `forceCloseFileChannel()` in ReferenceCountedFileChannel
- deallocate() call this `forceCloseFileChannel()`
- in here, you just remove it from `logid2FileChannel` and ``forceCloseFileChannel()`  it.

```
ReferenceCountedFileChannel fileChannel = logid2FileChannel.remove(logId);
if (null != fileChannel) {
    fileChannel.forceCloseFileChannel();
}
```